source/common/memory/stats.cc

```diff
@@ -2,121 +2,153 @@
   return value;
+#else
+  return 0;
+#endif
 }
 
 uint64_t Stats::totalThreadCacheBytes() {
+#if defined(TCMALLOC)
+  return tcmalloc::MallocExtension::GetNumericProperty("tcmalloc.current_total_thread_cache_bytes")
+      .value_or(0);
+#elif defined(GPERFTOOLS_TCMALLOC)
   size_t value = 0;
   MallocExtension::instance()->GetNumericProperty("tcmalloc.current_total_thread_cache_bytes",
                                                   &value);
   return value;
+#else
+  return 0;
+#endif
 }
 
 uint64_t Stats::totalPageHeapFree() {
+#if defined(TCMALLOC)
+  return tcmalloc::MallocExtension::GetNumericProperty("tcmalloc.pageheap_free_bytes").value_or(0);
+#elif defined(GPERFTOOLS_TCMALLOC)
   size_t value = 0;
   MallocExtension::instance()->GetNumericProperty("tcmalloc.pageheap_free_bytes", &value);
   return value;
+#else
+  return 0;
+#endif
 }
 
 uint64_t Stats::totalPageHeapUnmapped() {
+#if defined(TCMALLOC)
+  return tcmalloc::MallocExtension::GetNumericProperty("tcmalloc.pageheap_unmapped_bytes")
+      .value_or(0);
+#elif defined(GPERFTOOLS_TCMALLOC)
   size_t value = 0;
   MallocExtension::instance()->GetNumericProperty("tcmalloc.pageheap_unmapped_bytes", &value);
   return value;
+#else
+  return 0;
+#endif
 }
 
 uint64_t Stats::totalPhysicalBytes() {
+#if defined(TCMALLOC)
+  return tcmalloc::MallocExtension::GetProperties()["generic.physical_memory_used"].value;
+#elif defined(GPERFTOOLS_TCMALLOC)
   size_t value = 0;
   MallocExtension::instance()->GetNumericProperty("generic.total_physical_bytes", &value);
   return value;
+#else
+  return 0;
+#endif
 }
 
 void Stats::dumpStatsToLog() {
+#if defined(TCMALLOC)
+  ENVOY_LOG_MISC(debug, "TCMalloc stats:\n{}", tcmalloc::MallocExtension::GetStats());
+#elif defined(GPERFTOOLS_TCMALLOC)
   constexpr int buffer_size = 100000;
   auto buffer = std::make_unique<char[]>(buffer_size);
   MallocExtension::instance()->GetStats(buffer.get(), buffer_size);
   ENVOY_LOG_MISC(debug, "TCMalloc stats:\n{}", buffer.get());
+#else
+  return;
+#endif
 }
 
-} // namespace Memory
-} // namespace Envoy
-
+void AllocatorManager::tcmallocRelease() {
+#if defined(TCMALLOC)
+  tcmalloc::MallocExtension::ReleaseMemoryToSystem(bytes_to_release_);
 #else
+  return;
+#endif
+}
 
-namespace Envoy {
-namespace Memory {
-
-uint64_t Stats::totalCurrentlyAllocated() { return 0; }
-uint64_t Stats::totalThreadCacheBytes() { return 0; }
-uint64_t Stats::totalCurrentlyReserved() { return 0; }
-uint64_t Stats::totalPageHeapUnmapped() { return 0; }
-uint64_t Stats::totalPageHeapFree() { return 0; }
-uint64_t Stats::totalPhysicalBytes() { return 0; }
-void Stats::dumpStatsToLog() {}
+/**
+ * Configures tcmalloc release rate from the page heap. If `bytes_to_release_`
+ * has been initialized to `0`, no heap memory will be released in background.
+ */
+void AllocatorManager::configureBackgroundMemoryRelease(Api::Api& api) {
+  RELEASE_ASSERT(!tcmalloc_thread_, "Invalid state, tcmalloc has already been initialised");
```

Comment:
I think we shouldn't has this thread spinning if it's just doing nothing -- either we should config our way out, or not start the thread

Reply:
So what will happen in this case?  It's good to avoid release asserts so if the process could limp along, ENVOY_BUG is prefered

Reply:
We could use `ENVOY_BUG` yes, just error in logs and skip memory releasing setup.